NewareNDA/NewareNDA.py

```diff
@@ -177,33 +186,70 @@ def _aux_bytes_to_list(bytes):
     return [Index, Aux, T/10]
 
 
-def _generate_cycle_number(df):
+def _generate_cycle_number(df, cycle_mode='chg'):
     """
-    Generate a cycle number to match Neware. A new cycle starts with a charge
-    step after there has previously been a discharge.
+    Generate a cycle number to match Neware.
+    
+    cycle_mode = chg: (Default) Sets new cycles with a Charge step following a Discharge.
+        dchg: Sets new cycles with a Discharge step following a Charge.
+        auto: Identifies the first non-rest state as the incremental state.
     """
-
-    # Identify the beginning of charge steps
-    chg = (df['Status'] == 'CCCV_Chg') | (df['Status'] == 'CC_Chg') |  (df['Status'] == 'CP_Chg')
-    chg = (chg - chg.shift()).clip(0)
-    chg.iat[0] = 1
+    
+    # Auto: find the first non rest cycle
+    if cycle_mode.lower() == 'auto':
+        first_state = df[df['Status'] != 'Rest']['Status'].iat[0]
+        try:
+            _, cycle_mode = first_state.split('_', 1)
+        except ValueError:
+            # Status is SIM or otherwise. Set mode to chg
+            warnings.warn("First Step not recognized. Defaulting to Cycle_Mode 'Charge'.")
+            cycle_mode = 'chg'
+        
+    # Set increment key and non-increment/off key    
+    if cycle_mode.lower() == 'chg':
+        inkey = 'Chg'
+        offkey = 'DChg'
+    elif cycle_mode.lower() == 'dchg':
+        inkey = 'DChg'
+        offkey = 'Chg'
+    else: raise KeyError("Cycle_Mode string not recognized. Supported options are 'chg', 'dchg', and 'auto'.")

```

Comment:
Perhaps print erroneous mode in error, like:

```py
raise KeyError("Cycle_Mode {} not recognized. Supported options are 'chg', 'dchg', and 'auto'.".format(cycle_mode))
```

Reply:
Good idea. f-strings would also work:
`raise KeyError(f"Cycle_Mode {cycle_mode} not recognized. Supported options are 'chg', 'dchg', and 'auto'.")`